### ISSUE TITLE
[Reproducibility] Calling a pipeline with the exact same inputs yields different results

### ISSUE THREAD

**Issue by sayakpaul** [huggingface/diffusers#4055]
If I do:

```python
from diffusers import DiffusionPipeline
import numpy as np
import torch 

generator = torch.manual_seed(0)
dtype = torch.float16

pipeline = DiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5",
    torch_dtype=dtype,
    safety_checker=None
).to("cuda")

latents = pipeline("hello", num_inference_steps=2, generator=generator, output_type="np").images[0]

latents_2 = pipeline("hello", num_inference_steps=2, generator=generator, output_type="np").images[0]

print(np.allclose(
    latents[-3:, -3:, -1],
    latents_2[-3:, -3:, -1],
    atol=1e-3
))
```

It yields False. Shouldn't the outputs be the same?

**Comment by sayakpaul** [comment]
Thanks. That solves it!

**Comment by bghira** [comment]
for anyone else who stumbles upon this: the generator contains a set of _**random states**_ and those are updated as the inference runs. it's not necessarily just used to seed the random states, which is counter-intuitive.
